torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
     // TODO: only run while the last iteration has made a change
     size_t num_optimization_iters = 6;
     for (size_t i = 0; i < num_optimization_iters; i++) {
-      substituteInputTensorProperties();
+      // XXX: we cannot substitute symbolic dims before passes like constant
+      // propagation, or we might inadvertently use them in arithmetic or
+      // other operators
+      substituteInputTensorProperties(/*substitute_symbolic_dims*/ false);
       LowerSimpleTuples(graph_);
       RemoveListMutation(graph_);
       UnrollConstantLoops(graph_);
       ConstantPropagation(graph_);
       PeepholeOptimize(graph_);
       ConstantPropagation(graph_);
+      EliminateCommonSubexpression(graph_);
     }
-    ConstantPooling(graph_);
-    EliminateDeadCode(graph_);
+    substituteInputTensorProperties(/*substitute_symbolic_dims*/ true);
+    // XXX: do not run any passes aftr we have substituted in symbolic dimension
+    // value, we do it so they can be easily extracted into the output shape
     return extractOutputShape();
   }
 
  private:
-  void substituteInputTensorProperties() {
-    for (auto index : node_input_tensor_indices) {
-      substituteTensorProperties(index);
+  void substituteInputTensorProperties(bool substitute_symbolic_dims) {
```

Comment:
What is the difference between `substitute_symbolic_dims` being true or false? Adding a comment would help